### PR TITLE
[Quant][core][improvement] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case

### PR DESCRIPTION
**Pull request by dzdang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71269

Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1121,6 +1139,36 @@ static Tensor select_sparse(const Tensor& self, int64_t dim, int64_t index) {
...
+  }
+  QuantizerPtr quantizer;
+  auto temp = static_cast<PerChannelAffineQuantizer*>(quantizer_prev.get());
+  auto axis = temp->axis();
+  auto scales = temp->scales();
+  auto zero_points = temp->zero_points();
+  if (dim == axis) {
+    // Compute scales&zps for sub-tensor
+    is_select ? scales = scales.select(0, start) : scales = scales.slice(0, start, end, step);
+    is_select ? zero_points = zero_points.select(0, start) : zero_points = zero_points.slice(0, start, end, step);
```
Comment (jerryzh168) [inline aten/src/ATen/native/TensorShape.cpp:1158]: can select be expressed with slice?

Comment (jerryzh168) [reply aten/src/ATen/native/TensorShape.cpp:1158]: also, probably clearer if we do:
```
scales = is_select ? ... : ...;
```
if we need to do this

File: aten/src/ATen/native/quantized/Copy.cpp [hunk 2]
```diff
@@ -22,11 +27,19 @@ Tensor& quantized_copy_from_float_cpu_(Tensor& self, const Tensor& src) {
...
+                                         self.q_per_channel_zero_points(),
+                                         self.q_per_channel_axis());
+    } else {
+      float* src_data = src.data_ptr<float>();
+      scalar_t* self_data = self.data_ptr<scalar_t>();
+      for (const auto i : c10::irange(self.numel())) {
+        auto sc = self.q_scale();
+        auto test = self.q_zero_point();
+        self_data[i] = quantize_val<scalar_t>(
+            self.q_scale(), self.q_zero_point(), src_data[i]);
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/Copy.cpp:41]: can we call the quantize_tensor_per_tensor_affine here as well? (can be a separate PR)

File: test/quantization/core/test_quantized_tensor.py [hunk 3]
```diff
@@ -257,6 +257,45 @@ def _test_qtensor_sub_byte(self, rows, cols, dtype, elements_per_byte, expected_
...
+                # 3D tensor assignment verification
+                qt1[i][j:] = t2[i][j:]
+                self.assertEqual(qt1[i][j:], qt2[i][j:])
+                qt1[i] = t2[i]
+                self.assertEqual(qt1[i], qt2[i])
+                # 4D tensor assignment verification
+                qt1[:1] = t2[:1]
+                self.assertEqual(qt1[:1], qt2[:1])
+                qt1[:] = t2[:]
+                self.assertEqual(qt1[:], qt2[:])
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_tensor.py:297]: should we add a test for non-contiguous tensor slicing as well? 
we can use assertRaises: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L274

**Review by jerryzh168 — APPROVED** [review]
Looks great, thanks for all the detailed comments and tests!

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (dzdang) [reply aten/src/ATen/native/TensorShape.cpp:1158]: > can select be expressed with slice?

i think so, let me think. slice would maintain the dimension of the scales&zp to be 1 but select would just make it a scalar but that shouldn't affect the rest of the code

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).
